ttnn/cpp/ttnn/tensor/types.hpp

```diff
@@ -407,206 +396,220 @@ struct BorrowedStorage {
+        ordered_device_ids = other.ordered_device_ids;
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
+    MultiDeviceStorage(const MultiDeviceStorage &other) {
+        std::scoped_lock buf_lock(buffer_mtx, shape_mtx);
+        ordered_device_ids = other.ordered_device_ids;
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
 
-        MultiDeviceStorage &operator=( MultiDeviceStorage &&other) {
-            std::scoped_lock buf_lock(buffer_mtx, shape_mtx);
-            ordered_device_ids = other.ordered_device_ids;
-            strategy = other.strategy;
-            buffers = other.buffers;
-            shapes = other.shapes;
-            return *this;
-        }
+    MultiDeviceStorage &operator=(const MultiDeviceStorage &other) {
+        std::scoped_lock buf_lock(buffer_mtx, shape_mtx);
+        ordered_device_ids = other.ordered_device_ids;
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
 
-        bool operator == (const MultiDeviceStorage& other) {
-            return this->ordered_device_ids == other.ordered_device_ids and this->strategy == other.strategy and this->buffers == other.buffers and this->shapes == other.shapes;
-        }
+    MultiDeviceStorage &operator=(MultiDeviceStorage &&other) {
+        std::scoped_lock buf_lock(buffer_mtx, shape_mtx);
+        ordered_device_ids = other.ordered_device_ids;
+        strategy = other.strategy;
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
 
-        inline const MemoryConfig memory_config() const {
-            std::lock_guard<std::mutex> lock(buffer_mtx);
-            auto first_device_id = this->ordered_device_ids.at(0);
-            if (this->buffers.at(first_device_id).get() == nullptr) {
-                TT_THROW("MemoryConfig can only be obtained if the buffer is not null");
-            }
-            std::optional<ShardSpec> shard_spec = std::nullopt;
-            if (is_sharded(this->buffers.at(first_device_id)->buffer_layout())) {
-                shard_spec = this->buffers.at(first_device_id)->shard_spec().tensor_shard_spec;
-            }
-            return MemoryConfig{
-                .memory_layout = this->buffers.at(first_device_id)->buffer_layout(),
-                .buffer_type = this->buffers.at(first_device_id)->buffer_type(),
-                .shard_spec = shard_spec};
+    bool operator==(const MultiDeviceStorage &other) {
+        return this->ordered_device_ids == other.ordered_device_ids and this->strategy == other.strategy and
+               this->buffers == other.buffers and this->shapes == other.shapes;
+    }
 
+    inline const MemoryConfig memory_config() const {
+        std::lock_guard<std::mutex> lock(buffer_mtx);
+        auto first_device_id = this->ordered_device_ids.at(0);
```

Comment:
overall I prefer do manual check
if (this->ordered_device_ids.size() == 0) {
TT_FATAL("no such device...")
}
auto first_device_id = this->ordered_device_ids[0];
In this case we throw much better error message.